Us 15.3 outdoor n navigation strategy

JeremyVieira opened the pull request:
**Description**
1. Broke down outdoor navigation into individual steps
2. Created visual component to show indoor and outdoor navigation steps
3. Supports multiple campus buildings (EV, MB, VE, H, etc...)
4. Combines indoor/room to room navigation in parent branch (US15) to display any combination of indoor and outdoor navigation


**How was this tested?**
- [x]  Manual Test:
- [x]  Unit Test:


https://github.com/user-attachments/assets/ac8d87cd-6dc1-4962-966b-b5aa5de484b3

JeremyVieira commented:
Trying to fix the sonarqube reliability issues, once theyre all done, I will write more test cases to increase code coverage

JeremyVieira commented:
Reliability issues are fixed, will have to work on adding test cases for new code

IdrisDrouiche commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -373,16 +383,114 @@ const RoomToRoomNavigation = () => {
       );
 
+      // Map 'entrance' to an actual node in the graph if needed
+      let actualStartRoom = selectedStartRoom;
```
Consider renaming `actualStartRoom` to something more domain-specific like `mappedStartRoom` or `resolvedStartRoom` to clarify its role. This follows the Rename Variable pattern for better readability and domain alignment.

IdrisDrouiche commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -373,16 +383,114 @@ const RoomToRoomNavigation = () => {
+      }
+
+      // Also check if the end room exists
+      if (!Object.keys(endFloorGraph).includes(selectedEndRoom)) {
```
You could optionally abstract these checks into a helper like `validateNodeExists(graph, node)` for reusability and semantic clarity.

IdrisDrouiche commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -373,16 +383,114 @@ const RoomToRoomNavigation = () => {
+      console.log("Using end room:", selectedEndRoom);
+
+      // Make sure the mapped nodes actually exist in the graph
+      if (!Object.keys(startFloorGraph).includes(actualStartRoom)) {
```
same thing as previous comment you could use a `validateNodeExists(graph, node)` helper function

IdrisDrouiche reviewed (changes requested):
Wow Jeremy amazing work! I just left a few review comments mainly regarding refactoring and some suggestions on how to improve the quality of the code. When you have time take a look at the feedback. 

Apart from that great job brother 👊.

GitExcited commented on ConcordiaMaps/components/LiveBusTracker.js:
```diff
@@ -55,29 +68,49 @@ const LiveBusTracker = () => {
+  errorText: {
+    color: "white",
+    textAlign: "center",
+  },
```
These styles should be on its own file following the extract Variable refactoring

JeremyVieira replied:
Moved LiveBusTracker to its own styles file

JeremyVieira commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -71,14 +71,18 @@ const RoomToRoomNavigation = () => {
       );
 
+      // Use the passed buildingType from params if available, otherwise determine it
+      const determinedBuildingType =
```
This uses the findBuildingTypeFromId function found within NavigationValidationService file already

JeremyVieira commented:
Refactored the code to move functions outside of the MultiStepNavigationScreen and to use already existing functions. Moved building validation/ normalizing room ids to the BuildingDataService file. Used google maps outdoor directions from the useGoogleMapDirections hook. Also removed duplicate generateFloorPlanHtml and used the one inside FloorPlanService file.  Ultimately there was a lot of duplicate code and tried to reuse code as much as possible or move functions to relevant components

mannrp commented:
Thank you for pointing out the potential abstraction there and the renames. 

As suggested, 
- RoomToRoomNavigation was refactored and a new utility service was extracted to serve it.
- The new utility is tested as well.
- Variables holding room names were renamed to improve clarity.

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  const mapWebViewRef = useRef(null);
+
+  // Generate a random session token for Google Places API
+  const generateRandomToken = async () => {
```
Could this be place in its own helper file? For functions like generateRandomTokens. This would work as a refactoring for Extract Function

JeremyVieira replied:
Just pushed a change to move the generateRandomToken to the useGoogleMaps hook file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+    }
+  }, [route.params]);
+
+  const fetchOutdoorDirections = async (step) => {
```
We have a file for hooks which contains a googleMaps hook for directions. Would this fit better there?

JeremyVieira replied:
Pushed a change too put that method in the useGoogleMaps hook

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  }, []);
+
+  // Search for places with Google Places API
+  const searchOriginPlaces = async (text) => {
```
Same as this, these type of functions seem too all use google Maps API. You can create a hook or service to increase **maintainability** and **readability** as well as **reusability**

JeremyVieira replied:
Moved searchOriginPlaces and searchDestinationPlaces to hooks file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
-      <Header />
-      <NavBar />
+  // Search destination places with Google Places API
+  const searchDestinationPlaces = async (text) => {
```
Another Google place API call which would fit better on its own file

JeremyVieira replied:
Moved searchOriginPlaces and searchDestinationPlaces to hooks file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Handle destination selection from autocomplete
+  const handleDestinationSelection = async (placeId, description) => {
```
Another google maps API call which would fit better on its own file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Filter building suggestions based on text input
+  const filterBuildingSuggestions = (text) => {
```
We have a BuildingDataService file which is supposed to the the source of knowledge regarding Concordia Buildings. Following the principles of refactoring and single responsability principle, this should be in there.

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Handle building suggestion filtering
+  const filterOriginBuildingSuggestions = (text) => {
```
This is another opperation on buildings that could be better on BuildingDataService or on its own file that takes care of BuildingSuggestions based on text

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+    setIsLoading(true);
+
+    // Create navigation steps
+    const steps = [];
```
It is very important that this navigation steps logic is extracted from this code into its own component. For maintainability purposes, the steps creating business logic should be decoupled  from the Screen UI component MultiStepNavigation

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Generate Google Maps HTML for outdoor steps
+  const generateMapHtml = () => {
```
The outdoor Steps should be its own component for maintanability and reusability. The MultiStepNavigation should import this component and pass the approprite props

JeremyVieira replied:
Moved the generateMapHTMl to its own services file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Render indoor navigation UI with floor plans and step-by-step directions
+  const renderIndoorNavigation = () => {
```
Same thing for indoorNavigation, it should be its own component, separate from the outdoornavigation UI logic as well as the multistepnavigation logic

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Render indoor step UI with button to open RoomToRoomNavigation
+  const renderIndoorStep = () => {
```
This should be its own component too

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Render navigation form if no plan is active
+  const renderNavigationForm = () => {
```
This ia another great example of a component. These are all examples of the refactoring Extract Class

JeremyVieira replied:
This file is meant to be the UI screen. The function generates the layout of what is meant to be displayed. Should it be defined in a different component and just called in this file when rendering the UI component?

GitExcited replied:
Yes it should be defined in another file. Following the principle of single-concern, a jsx file should only return one component. If you find yourself creating mulitplle returns, its likely they can be placed in their own file. This ensures maintainability, testability and redability.

JeremyVieira replied:
Moved the form generation to its own component

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Render outdoor step UI with map and directions
+  const renderOutdoorStep = () => {
```
Another UI component that can be extracted

JeremyVieira replied:
Moved it to NavigationStep file

GitExcited commented on ConcordiaMaps/components/MultistepNavigation/MultistepNavigationScreen.js:
```diff
@@ -1,141 +1,2114 @@
+  };
+
+  // Render active navigation steps UI
+  const renderNavigationSteps = () => {
```
Another UI Component that can be extracted

JeremyVieira replied:
Now uses the NavigationStep component to show the navigation steps

GitExcited reviewed (changes requested):
I reviewed all the remaining opportunities for further refactoring. It mainly concerns extracting functions ( Google API calls, BuildingService data ) as well as decoupling UI components using the Extract Class refactoring. This ensures each file follows the single-concern principle and that the code is maintainable, reusable and readable.

shanvinluo commented on ConcordiaMaps/__tests__/components/MultistepNavigation/MultistepNavigationScreen.test.js:
Just wanted to comment on the impressive test coverage. Great job!

shanvinluo commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -373,16 +388,102 @@ const RoomToRoomNavigation = () => {
+
+      // Find a suitable entrance node based on the building type and available nodes
+      if (selectedStartRoom === "entrance") {
+        if (buildingType === "HallBuilding" && startFloor === "1") {
```
The values "HallBuilding", "1", and "entrance" were hardcoded in multiple places. Extracting them into constants (BUILDINGS.HALL, FIRST_FLOOR, DEFAULT_ENTRANCE) improves maintainability and reduces the risk of typos. This follows the ["Replace Magic Literal"](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA) refactoring from the catalog.

shanvinluo commented on ConcordiaMaps/components/IndoorNavigation/RoomToRoomNavigation.js:
```diff
@@ -373,16 +388,102 @@ const RoomToRoomNavigation = () => {
+      let resolvedStartRoom = selectedStartRoom;
+
+      // Find a suitable entrance node based on the building type and available nodes
+      if (selectedStartRoom === "entrance") {
```
The entrance node selection logic is duplicated in multiple places. Extracting it into the resolveEntranceNode function improves reusability and reduces redundancy. This follows the ["Extract Function"](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA) refactoring from the catalog.

shanvinluo reviewed (changes requested):
Overall great job with the implementation, really impressed with what you achieved and even more impressive is the test coverage. The only problem with it are some light refactoring problems that e should resolve:

As for testing, I tested this feature both manually and through unit tests.

- I navigated through different buildings on the campus, such as EV, MB, VE, and H, to ensure that both indoor and outdoor navigation steps are properly displayed.
- I confirmed that the visual component correctly represents both indoor (room-to-room) and outdoor navigation steps.
- I tested various combinations of indoor and outdoor steps in the navigation flow (e.g., combining room-to-room navigation with building-to-building outdoor navigation).
- I verified that the navigation process works smoothly for different buildings and step sequences.

MomoG221 commented:
Great job, everyone! Could you please address the 32 issues related to the maintainability of the code, as highlighted by SonarQube? You can find them here: [SonarQube Issues](https://sonarcloud.io/project/issues?impactSoftwareQualities=MAINTAINABILITY&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=200&id=SOEN-390-pmzone_SOEN-_390-Q-QD).

JeremyVieira commented:
Refactored some tests and moved them to the hooks file. Seems like it slightly decreased the code coverage as well. That is why the checks failed

JeremyVieira commented:
Once again failed due to insufficient test cases (79.9% coverage)

JeremyVieira commented:
Prettier failed to pass check

JeremyVieira commented:
Failing due to insufficient test coverage

JeremyVieira commented:
Refactoring the test files decreased coverage, adding more test cases

GitExcited commented on ConcordiaMaps/services/NavigationPlanService.js:
```diff
@@ -0,0 +1,345 @@
+ * @param {Object} params.navigation - React Navigation object
+ */
+const createNavigationPlan = ({
+  originInputType,
```
This is a great example of [Extract Class](https://refactoring.com/catalog/extractClass.html) refactoring. Well done!

JeremyVieira commented:
Good work! Tested everything and it is working, successfully able to get different direction types and supports the different buildings. Refactoring done makes the code easier to read and understand

GitExcited commented:
After reviewing the remaining of the code I didn't identify any other major refactoring opportunities. Furthermore, the code passes all tests and quality checks from Sonarqube. This U.S. is ready to be merged

Atai-Askarov reviewed (approved):
Haven't noticed any opportunities for further refactoring. The PR is ready to be merged!

GitExcited reviewed (approved):
After reviewing the remaining of the code I didn't identify any other major refactoring opportunities. Furthermore, the code passes all tests and quality checks from Sonarqube. This U.S. is ready to be merged

shanvinluo reviewed (commented):
Forgot to approve it but LGTM! All the refactoring was addressed
